system/lib/pthread/library_pthread.c

```diff
@@ -0,0 +1,543 @@
+#define _GNU_SOURCE
+#include <pthread.h>
+#include <emscripten/threading.h>
+#include <emscripten.h>
+#include <sys/time.h>
+#include <dirent.h>
+#include <utime.h>
+#include <sys/stat.h>
+#include <sys/statvfs.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <poll.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <termios.h>
+#include <sys/mman.h>
+#include <sys/socket.h>
+#include <sys/ioctl.h>
+#include "../internal/pthread_impl.h"
+#include <assert.h>
+
+int _pthread_getcanceltype()
+{
+	return pthread_self()->cancelasync;
+}
+
+static void inline __pthread_mutex_locked(pthread_mutex_t *mutex)
+{
+	// The lock is now ours, mark this thread as the owner of this lock.
+	assert(mutex);
+	assert(mutex->_m_lock == 0);
+	mutex->_m_lock = pthread_self()->tid;
+	if (_pthread_getcanceltype() == PTHREAD_CANCEL_ASYNCHRONOUS) pthread_testcancel();
+}
+
+double _pthread_nsecs_until(const struct timespec *restrict at)
+{
+	struct timeval t;
+	gettimeofday(&t, NULL);
+	double cur_t = t.tv_sec * 1e9 + t.tv_usec * 1e3;
+	double at_t = at->tv_sec * 1e9 + at->tv_nsec;
+	double nsecs = at_t - cur_t;
+	return nsecs;
+}
+
+#if 0
```

Comment:
I've gone back and forth a few times between this implementation for mutexes and the musl one, and I find value in this code as a potential alternative to the musl code, so I thought to keep it at hand for reference for now at least.